polars/polars-core/src/frame/row/transpose.rs

```diff
@@ -54,25 +61,56 @@ impl DataFrame {
                 }
                 let cols = buffers
                     .into_iter()
-                    .enumerate()
-                    .map(|(i, buf)| {
-                        let mut s = buf.into_series().cast(dtype).unwrap();
-                        s.rename(&format!("column_{i}"));
-                        s
-                    })
+                    .map(|buf| buf.into_series().cast(dtype).unwrap())
                     .collect::<Vec<_>>();
                 Ok(DataFrame::new_no_checks(cols))
             }
+        }?;
+        out.set_column_names(colnames)?;
+        match keep_names_as {
+            Some(cn) => {
+                let namecol = Utf8Chunked::new(cn, self.get_column_names());
+                out.insert_at_idx_no_name_check(0, namecol.into()).cloned()
+            }
+            None => Ok(out),
         }
     }
 
     /// Transpose a DataFrame. This is a very expensive operation.
-    pub fn transpose(&self) -> PolarsResult<DataFrame> {
+    pub fn transpose(
+        &self,
+        keep_names_as: Option<&str>,
+        column_names: Option<Either<String, Vec<String>>>,
+    ) -> PolarsResult<DataFrame> {
+        let mut df = self.clone(); // Must be owned so we get the same type if dropping a column.
+        let colnames_t = match column_names {
+            None => (0..self.height()).map(|i| format!("column_{i}")).collect(),
+            Some(cn) => match cn {
+                Either::Left(cname) => {
+                    let new_names = self.column(&cname).and_then(|x| x.utf8())?;
+                    polars_ensure!(!new_names.has_validity(), ComputeError: "Column with new names can't have null values");
+                    df = self.drop(&cname)?;
+                    new_names
```

Comment:
If we have `new_names ` scoped above, we can collect into a `Vec<&str>` and we don't need to heap allocate the strings.

Reply:
How do I make that work with the other branch that generates names by formatting? I couldn't figure out how to end up with &str instead of String because of that.

Reply:
Ah, right. Yeah, `from_dtypes` accepts `&[String]`. Let's leave that one for now.